---
title: "L.panamensis 202102: Differential Expression in human PBMCs followed by macrophages"
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    code_download: true
    code_folding: show
    df_print: paged
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    width: 300
    keep_md: false
    mode: selfcontained
    toc_float: true
  BiocStyle::html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    keep_md: false
    mode: selfcontained
    toc_float: true
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
 font-size: 16px
}
</style>

```{r options, include=FALSE}
library("hpgltools")
tt <- devtools::load_all("~/hpgltools")
knitr::opts_knit$set(progress=TRUE,
                     verbose=TRUE,
                     width=90,
                     echo=TRUE)
knitr::opts_chunk$set(error=TRUE,
                      fig.width=8,
                      fig.height=8,
                      dpi=96)
old_options <- options(digits=4,
                       stringsAsFactors=FALSE,
                       knitr.duplicate.label="allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size=10))
rundate <- format(Sys.Date(), format="%Y%m%d")
previous_file <- "index.Rmd"
ver <- "202102"

##tmp <- sm(loadme(filename=paste0(gsub(pattern="\\.Rmd", replace="", x=previous_file), "-v", ver, ".rda.xz")))
rmd_file <- "pbmc_all_analyses_202102.Rmd"
```

# Annotation version: `r ver`

## Genome annotation input

There are a few methods of importing annotation data into R.  The following are
two attempts, the second is currently being used in these analyses.

```{r make_orgdb_meta}
meta <- EuPathDB::download_eupath_metadata(webservice="tritrypdb", eu_version=46)

lm_entry <- EuPathDB::get_eupath_entry(species="Leishmania major", metadata=meta)
lp_entry <- EuPathDB::get_eupath_entry(species="Leishmania panamensis", metadata=meta)
lmex_entry <- EuPathDB::get_eupath_entry(species="Leishmania mexicana", metadata=meta)
lama_entry <- EuPathDB::get_eupath_entry(species="Leishmania amazonensis", metadata=meta)
lb_entry <- EuPathDB::get_eupath_entry(species="2904", metadata=meta)
ld_entry <- EuPathDB::get_eupath_entry(species="donovani", metadata=meta)
crit_entry <- EuPathDB::get_eupath_entry(species="Crith", metadata=meta)
```

```{r make_orgdb, eval=FALSE}
testing_panamensis <- EuPathDB::make_eupath_orgdb(entry=lp_entry)
testing_braziliensis <- EuPathDB::make_eupath_orgdb(entry=lb_entry)
testing_donovani <- EuPathDB::make_eupath_orgdb(entry=ld_entry)
testing_mexicana <- EuPathDB::make_eupath_orgdb(entry=lmex_entry)
testing_major <- EuPathDB::make_eupath_orgdb(entry=lm_entry)
testing_crith <- EuPathDB::make_eupath_orgdb(entry=crit_entry)
```

Assuming the above packages got created, we may load them and extract the
annotation data.

```{r lpanamensis_orgdb}
wanted_fields <- c("annot_cds_length", "annot_chromosome", "annot_gene_entrez_id",
                   "annot_gene_name", "annot_strand", "gid", "go_go_id",
                   "go_go_term_name", "go_ontology",
                   "interpro_description" ,"interpro_e_value", "type_gene_type")

## The last set of these I created with the old EuPathDB was version 46
## I also have new ones for version 49, but figured I should stick with the old ones for now.
lm_org <- sm(EuPathDB::load_eupath_annotations(query=lm_entry, eu_version=46))
lp_org <- sm(EuPathDB::load_eupath_annotations(query=lp_entry, eu_version=46))
lb_org <- sm(EuPathDB::load_eupath_annotations(query=lb_entry, eu_version=46))
ld_org <- sm(EuPathDB::load_eupath_annotations(query=ld_entry, eu_version=46))
lmex_org <- sm(EuPathDB::load_eupath_annotations(query=lmex_entry, eu_version=46))
cf_ort <- sm(EuPathDB::load_eupath_annotations(query=crit_entry, eu_version=46))
```

## Read a gff file

In contrast, it is possible to load most annotations of interest directly from
the gff files used in the alignments.  More in-depth information for the human
transcriptome may be extracted from biomart.

```{r genome_input}
## The old way of getting genome/annotation data
lp_gff <- "reference/lpanamensis.gff"
lb_gff <- "reference/lbraziliensis.gff"
hs_gff <- "reference/hsapiens.gtf"

lp_fasta <- "reference/lpanamensis.fasta.xz"
lb_fasta <- "reference/lbraziliensis.fasta.xz"
hs_fasta <- "reference/hsapiens.fasta.xz"

lp_annotations <- sm(load_gff_annotations(lp_gff, type="gene"))
rownames(lp_annotations) <- paste0("exon_", lp_annotations$web_id, ".1")

lb_annotations <- sm(load_gff_annotations(lb_gff, type="gene"))

hs_gff_annot <- sm(load_gff_annotations(hs_gff, id_col="gene_id"))
hs_annotations <- sm(load_biomart_annotations())$annotation
hs_annotations$ID <- hs_annotations$geneID
rownames(hs_annotations) <- make.names(hs_annotations[["ensembl_gene_id"]], unique=TRUE)
dim(hs_annotations)

lp_size_dist <- plot_histogram(lp_annotations[["width"]])
lp_size_dist

hs_size_dist <- plot_histogram(hs_annotations[["cds_length"]])
hs_size_dist +
  ggplot2::scale_x_continuous(limits=c(0, 20000))
```

## Getting ontology data

Annotation for gene ontologies may be gathered from a similarly large number of
sources. The following are a couple.

```{r ontology1}
## Try using biomart
hs_go <- sm(load_biomart_go())
## or the org.Hs.eg.db sqlite database
tt <- sm(library("Homo.sapiens"))
hs <- Homo.sapiens
##hs_go_ensembl <- load_orgdb_go(hs, hs_annotations$geneID)
##dim(hs_go_biomart)
##dim(hs_go_ensembl)
##hs_goids <- hs_go_biomart


## While testing, I called this desc, that will need to change.
##lp_tooltips <- make_tooltips(lp_annotations)
##lb_tooltips <- make_tooltips(lb_annotations)

lp_lengths <- lp_annotations[, c("ID", "width")]
lb_lengths <- lb_annotations[, c("ID", "width")]
hs_lengths <- hs_annotations[, c("ensembl_gene_id", "cds_length")]
colnames(hs_lengths) <- c("ID", "width")

lp_goids <- read.csv(file="reference/lpan_go.txt.xz", sep="\t", header=FALSE)
lb_goids <- read.csv(file="reference/lbraz_go.txt.xz", sep="\t", header=FALSE)
colnames(lp_goids) <- c("ID","GO","ont","name","source","tag")
colnames(lb_goids) <- c("ID","GO","ont","name","source","tag")
```

# Putting the pieces together

The PBMC experiment has samples across 2 contexts, the host and parasite.
The following block sets up one experiment for each.  If you open the
all_samples-species.xlsx files, you will note immediately that a few different
attempts were made at ascertaining the most likely experimental factors that
contributed to the readily apparent batch effects.

Start out by extracting the relevant data and querying it to see the general quality.

This first block sets the names of the samples and colors.
It also makes separate data sets for:

* All features with infected and uninfected samples.
* All features with only the infected samples.
* Only CDS features with infected and uninfected samples.
* Only CDS features with only the infected samples.

```{r infection_expt}
hs_final_annotations <- hs_annotations
hs_final_annotations <- hs_final_annotations[, c("ensembl_transcript_id", "ensembl_gene_id", "cds_length",
                                                 "hgnc_symbol", "description", "gene_biotype")]
hs_final_annotations$rn <- rownames(hs_final_annotations)
note <- "New experimental design factors by snp added 2016-09-20"
rownames(hs_final_annotations) <- hs_final_annotations$rn
hs_final_annotations$rn <- NULL

hs_length <- hs_final_annotations[, c("ensembl_gene_id", "cds_length")]
colnames(hs_length) <- c("ID", "length")

hs_pbmc <- create_expt(
    metadata=glue::glue("sample_sheets/pbmc_samples_{ver}.xlsx"),
    gene_info=hs_final_annotations,
    file_column="humanfile",
    notes=note)
hs_inf <- create_expt(
    metadata=glue::glue("sample_sheets/pbmc_samples_removetwo_{ver}.xlsx"),
    gene_info=hs_final_annotations,
    file_column="humanfile",
    notes=note)
chosen_colors <- c("#009900","#990000", "#000099")
names(chosen_colors) <- c("uninf","chr","sh")
hs_pbmc <- set_expt_colors(hs_pbmc, colors=chosen_colors)
hs_inf <- set_expt_colors(hs_inf, colors=chosen_colors)

inf_newnames <- paste0(pData(hs_inf)$label, "_", pData(hs_inf)$donor)
hs_inf <- set_expt_samplenames(hs_inf, newnames=inf_newnames)

hs_cds_pbmc <- exclude_genes_expt(hs_pbmc, method="keep",
                                  column="gene_biotype",
                                  patterns="protein_coding")
hs_cds_inf <- exclude_genes_expt(hs_inf, method="keep",
                                 column="gene_biotype",
                                 patterns="protein_coding")
```

## The parasite transcriptome mappings

```{r parasite_expt}
lp_pbmc_all <- sm(create_expt(
    metadata=glue::glue("sample_sheets/pbmc_samples_{ver}.xlsx"),
    gene_info=lp_annotations, file_column="parasitefile"))
lp_pbmc_inf <- sm(create_expt(
    metadata=glue::glue("sample_sheets/pbmc_samples_removetwo_{ver}.xlsx"),
    gene_info=lp_annotations, file_column="parasitefile"))
```

# Sample Estimation: `r ver`

```{r inf_plots, fig.show="hide"}
hs_inf_met <- sm(graph_metrics(hs_cds_inf))
hs_inf_write <- write_expt(
    hs_cds_inf, norm="quant", convert="cpm",
    transform="log2", batch="svaseq",
    filter=TRUE,
    excel=glue::glue("excel/pbmc_only_infected-v{ver}.xlsx"))
```

# My favorite path for the data

1.  Show a scaled (no batch) PCA of samples including the uninfected.
  a.  #1 following sva
2.  Show #1 without the uninfected.
  a.  #2 followig sva
3.  Show #2 without the two weirdo samples.
4.  Show #3 following sva.

```{r my_path}
## Step 1, all samples (cds only) including uninfected samples.
pbmc_norm <- normalize_expt(hs_cds_pbmc, transform="log2", convert="cpm",
                            norm="quant", filter=TRUE)
plt <- plot_pca(pbmc_norm)$plot
pp(file=glue::glue("images/pbmc_step1-v{ver}.pdf"), image=plt)
plt

## Step 1a.
pbmc_nb <- normalize_expt(hs_cds_pbmc, convert="cpm",
                          filter=TRUE, batch="svaseq")
pbmc_nb <- normalize_expt(pbmc_nb, transform="log2")
plt <- plot_pca(pbmc_nb)$plot
pp(file=glue::glue("images/pbmc_step1a-v{ver}-d{rundate}.pdf"), image=plt)
plt

## Step 2, drop the uninfected samples and repeat.
hs_minus <- subset_expt(hs_cds_pbmc, subset="condition!='uninf'")
## minus_norm <- normalize_expt(hs_minus, transform="log2", convert="cpm", norm="quant", filter=TRUE)
minus_norm <- normalize_expt(hs_minus, convert="cpm", norm="quant", filter=TRUE)
minus_norm <- normalize_expt(minus_norm, transform="log2")
plt <- plot_pca(minus_norm)$plot
pp(file=glue::glue("images/pbmc_step2-v{ver}.pdf"), image=plt)
plt

## Step 2a
##minus_nb <- normalize_expt(hs_minus, transform="log2", convert="cpm", filter=TRUE, batch="svaseq")
minus_nb <- normalize_expt(hs_minus, convert="cpm", filter=TRUE, batch="svaseq")
minus_nb <- normalize_expt(minus_nb, transform="log2")
plt <- plot_pca(minus_nb)$plot
pp(file=glue::glue("images/pbmc_step2a-v{ver}.pdf"), image=plt)
plt

## Step 3, drop the problematic samples and repeat.
minus_minus <- subset_expt(hs_cds_inf, subset="condition!='uninf'")
minusminus_norm <- normalize_expt(minus_minus, transform="log2", convert="cpm",
                                  norm="quant", filter=TRUE)
plt <- plot_pca(minusminus_norm)$plot
pp(file=glue::glue("images/pbmc_step3-v{ver}.pdf"), image=plt)
plt

## Step 4, add sva to step 3.
minusminus_batch <- normalize_expt(minus_minus, transform="log2", convert="cpm",
                                   batch="svaseq", filter=TRUE)
plt <- plot_pca(minusminus_batch)$plot
pp(file=glue::glue("images/pbmc_step4-v{ver}.pdf"), image=plt)
plt
## An alternate version, where the transformation is explicitly last

minusminus_batch <- normalize_expt(minus_minus, convert="cpm", batch="svaseq", filter=TRUE)
minusminus_batch <- normalize_expt(minusminus_batch, transform="log2")
plt <- plot_pca(minusminus_batch)$plot
pp(file=glue::glue("images/pbmc_step4_separate-v{ver}.pdf"), image=plt)
plt

step4_with_uninf <- normalize_expt(hs_cds_inf, convert="cpm", batch="svaseq", filter=TRUE)
step4_with_uninf <- normalize_expt(step4_with_uninf, transform="log2")
plt <- plot_pca(step4_with_uninf)$plot
pp(file=glue::glue("step4_with_uninfected_separate-v{ver}.pdf"), image=plt)
plt
```

# Perform Multiple DE

1.  Perform DE of step 2a data.
2.  Perform DE of step 4 data.
3.  Venn diagram of up/down in the two sets showing significant similarities.

```{r multi_de_strains}
keepers <- list("sh_nil" = c("sh", "uninf"),
                "ch_nil" = c("chr", "uninf"),
                "ch_sh" = c("chr", "sh"))
levels(pData(hs_cds_pbmc)$condition)
nrow(pData(hs_cds_pbmc))
pbmc_de <- all_pairwise(hs_cds_pbmc, model_batch="svaseq", filter=TRUE)
pbmc_tables <- combine_de_tables(
  pbmc_de, keepers=keepers,
  excel=glue::glue("excel/pbmc_de_tables_svaseq-v{ver}-d{rundate}.xlsx"))
pbmc_sig <- extract_significant_genes(
  pbmc_tables, lfc=0.585,
  excel=glue::glue("excel/pbmc_sig_tables_svaseq-v{ver}-d{rundate}.xlsx"))

pbmc_de_batch <- all_pairwise(hs_cds_pbmc, model_batch=TRUE, filter=TRUE)
pbmc_batch_tables <- combine_de_tables(
  pbmc_de_batch, keepers=keepers,
  excel=glue::glue("excel/pbmc_de_tables_batch-v{ver}-d{rundate}.xlsx"))
pbmc_batch_sig <- extract_significant_genes(
  pbmc_batch_tables, lfc=0.585,
  excel=glue::glue("excel/pbmc_sig_tables_batch-v{ver}-d{rundate}.xlsx"))

levels(pData(hs_cds_inf)$condition)
nrow(pData(hs_cds_inf))
minus_de <- all_pairwise(hs_cds_inf, model_batch="svaseq", filter=TRUE)
minus_tables <- combine_de_tables(
  minus_de, keepers=keepers,
  excel=glue::glue("excel/pbmc_de_tables_nouninf_svaseq-v{ver}-d{rundate}.xlsx"))
minus_sig <- extract_significant_genes(
  minus_tables, lfc=0.585,
  excel=glue::glue("excel/pbmc_sig_tables_nouninf_svaseq-v{ver}-d{rundate}.xlsx"))

minus_de_batch <- all_pairwise(hs_cds_inf, model_batch=TRUE, filter=TRUE)
minus_tables_batch <- combine_de_tables(
    minus_de_batch, keepers=keepers,
    excel=glue::glue("excel/pbmc_de_tables_nouninf_batch-v{ver}-d{rundate}.xlsx"))
minus_sig_batch <- extract_significant_genes(
    minus_tables_batch, lfc=0.585,
    excel=glue::glue("excel/pbmc_sig_tables_nouninf_batch-v{ver}-d{rundate}.xlsx"))

similarities <- compare_de_results(pbmc_tables, minus_tables)
```

# Some tasks 202008

## Supplemental figure pictures for the pbmc data (3 donors)

1.  Library sizes, distance heatmap

```{r s1_figures}
pbmc_s1_libsize <- plot_libsize(hs_cds_pbmc)
pp(file="pictures/pbmc_s1_libsize.pdf", image=pbmc_s1_libsize$plot)

pbmc_heat_norm <- normalize_expt(hs_cds_pbmc, filter=TRUE, batch="svaseq", transform="log2")
pbmc_s1_heat <- plot_disheat(pbmc_heat_norm)
pp(file="pictures/pbmc_s1_disheat.pdf", image=pbmc_s1_heat)
```

## Try out upsetr with the following for the pbmc data

1.  Include both up and down for:
  a.  |logFC| >= 1
  b.  adjp <= 0.05
  c.  deseq results
2.  Categories are sh/nil-up, chr/nil-up, sh/nil-down, chr/nil-down

Therefore I am reasonably certain the data to use for this operation is from
'all_sig'.  When using upsetr, it expectes a dataframe where columns are the
categories (up_sh_vs_nil, down_sh_vs_nil, etc), rows are genes, and each element
is the number of times that element is true for the given [gene, category].  In
this case, it will be either 0 or 1 for everything.


```{r upsetr}
start <- data.frame(row.names=rownames(exprs(hs_cds_pbmc)))
start[["up_sh_vs_nil"]] <- 0
start[["down_sh_vs_nil"]] <- 0
start[["up_chr_vs_nil"]] <- 0
start[["down_chr_vs_nil"]] <- 0
idx <- rownames(pbmc_sig[["deseq"]][["ups"]][["sh_nil"]])
start[idx, "up_sh_vs_nil"] <- 1
idx <- rownames(pbmc_sig[["deseq"]][["downs"]][["sh_nil"]])
start[idx, "down_sh_vs_nil"] <- 1
idx <- rownames(pbmc_sig[["deseq"]][["ups"]][["ch_nil"]])
start[idx, "up_chr_vs_nil"] <- 1
idx <- rownames(pbmc_sig[["deseq"]][["downs"]][["ch_nil"]])
start[idx, "down_chr_vs_nil"] <- 1

unique_sh_up_idx <- start[["up_sh_vs_nil"]] == 1 & rowSums(start) == 1
unique_sh_up <- start[unique_sh_up_idx, ]

unique_sh_down_idx <- start[["down_sh_vs_nil"]] == 1 & rowSums(start) == 1
unique_sh_down <- start[unique_sh_down_idx, ]

test_up <- simple_gprofiler(sig_genes=rownames(unique_sh_up), species="hsapiens")
test_down <- simple_gprofiler(sig_genes=rownames(unique_sh_down), species="hsapiens")
```

## Use UpSetR to get an idea of how much overlap there is in these categories

Thus, in the following picture we see:

1. ch/nil going up has 128 unique genes.
2. sh/nil going up has 139 unique genes.
3. ch/nil going down has 156 unique genes.
4. sh/nil going down has 243 unique genes.
5. There are 1403 shared up genes.
6. There are 791 shared down genes.

```{r upsetr}
library(UpSetR)
plt <- upset(start)
plt <- grDevices::recordPlot()
pp(file="pictures/upset.pdf", image=plt)
```

# Deconvolution/GSVA analyses with these samples

Let us grab the C7 immunology signatures from msigdb, there are a few ways to do this.
One is to parse the xml files and another to parse the gmt files.

```{r msig_immunology}
broad_c7 <- GSEABase::getGmt("reference/msigdb_v7.2/c7.all.v7.2.entrez.gmt",
                             collectionType=GSEABase::BroadCollection(category="c7"),
                             geneIdType=GSEABase::EntrezIdentifier())

broad_c8 <- GSEABase::getGmt("reference/msigdb_v7.2/c8.all.v7.2.entrez.gmt",
                             collectionType=GSEABase::BroadCollection(category="c8"),
                             geneIdType=GSEABase::EntrezIdentifier())

broad_c2 <- GSEABase::getGmt("reference/msigdb_v7.2/c2.all.v7.2.entrez.gmt",
                             collectionType=GSEABase::BroadCollection(category="c2"),
                             geneIdType=GSEABase::EntrezIdentifier())
```

```{r gsva_reactome}
better_names <- pData(hs_cds_pbmc)[["label"]]
hs_cds_pbmc_new <- set_expt_samplenames(expt=hs_cds_pbmc, newnames=better_names)
hs_cds_gsva_input <- normalize_expt(hs_cds_pbmc_new, filter=TRUE, batch="svaseq")
hs_cds_gsva_input <- normalize_expt(hs_cds_gsva_input, convert="rpkm", column="cds_length")
pbmc_gsva <- simple_gsva(hs_cds_gsva_input)
gsva_expt <- pbmc_gsva[["expt"]]
pbmc_scored <- gsva_likelihoods(pbmc_gsva, factor="chr")

reactome_subset <- grepl(x=rownames(gsva_expt$expressionset), pattern="^REACTOME")
reactome_gsva <- gsva_expt$expressionset[reactome_subset, ]
pp("images/pbmc_reactome_gsva.svg")
color_range <- c("#00007F", "blue", "#007FFF", "cyan",
                 "#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000")
jet_colors <- grDevices::colorRampPalette(color_range)
tt <- heatmap.3(exprs(reactome_gsva), cexRow=0.1, cexCol=0.5, trace="none", col=jet_colors)
dev.off()
```

A couple of TODOs from our meeting 20210224

1.  Add some metadata to the gsva data from the msig xml file: organism, description_brief, authors.
2.  Perform the gsva limma of infected vs. uninfected.
3.  Incorporate Theresa's improvements

```{r todo_testing}
testing <- simple_gsva(hs_cds_gsva_input, signatures = broad_c7,
                       msig_xml = "reference/msigdb_v7.2.xml", cores = 10)
```

## C2 MSigDB with GSVA

Here is an invocation of all of my current analyses with gsva using the C2 signatures.

```{r gsva_c2_msig_categories}
c2_pbmc_gsva <- simple_gsva(hs_cds_gsva_input, signatures=broad_c2)
c2_pbmc_sig <- get_sig_gsva_categories(
    c2_pbmc_gsva,
    excel=glue::glue("excel/pbmc_batch_msig_c2_categories-v{ver}.xlsx"))

pp(file="images/pbmc_c2_gsva_scores.svg")
c2_pbmc_sig$raw_plot
dev.off()
pp(file="images/pbmc_c2_likelihoods.svg")
c2_pbmc_sig$likelihood_plot
dev.off()
pp(file="images/pbmc_c2_subset.svg")
c2_pbmc_sig$subset_plot
dev.off()
```

## C7 MSigDB with GSVA

Ibid, but this time with C7.

```{r gsva_c7_msig_categories}
c7_pbmc_gsva <- simple_gsva(hs_cds_gsva_input, signatures=broad_c7,
                            msig_xml = "reference/msigdb_v7.2.xml", cores = 10)
c7_pbmc_sig <- get_sig_gsva_categories(
    c7_pbmc_gsva,
    excel=glue::glue("excel/pbmc_msig_c7_categories-v{ver}.xlsx"))

pp(file="images/pbmc_c7_gsva_scores.svg")
c7_pbmc_sig$raw_plot
dev.off()
pp(file="images/pbmc_c7_likelihoods.svg")
c7_pbmc_sig$likelihood_plot
dev.off()
pp(file="images/pbmc_c7_subset.svg")
c7_pbmc_sig$subset_plot
dev.off()
```

# Use msigdb along with goseq and the up/down genes

```{r msig_goseq_up_chr}
ups <- pbmc_sig[["deseq"]][["ups"]][["ch_nil"]]
chr_uninf_msig_goseq <- goseq_msigdb(sig_genes=ups, signatures=broad_c7,
                                     signature_category="c7",
                                     length_db=hs_length,
                                     excel=glue::glue("excel/pbmc_chr_uninf_up_msig_goseq-v{ver}.xlsx"))
pp(file="images/pbmc_chronic_vs_uninfected_msig_goseq.png", width=18)
chr_uninf_msig_goseq$pvalue_plots$mfp_plot_over
dev.off()
downs <- pbmc_sig[["deseq"]][["downs"]][["ch_nil"]]
chr_uninf_down_msig_goseq <- goseq_msigdb(sig_genes=downs, signatures=broad_c7,
                                          signature_category="c7",
                                          length_db=hs_length,
                                          excel=glue::glue("excel/pbmc_chr_uninf_down_msig_goseq-v{ver}.xlsx"))
pp(file="images/pbmc_chronic_vs_uninfected_msig_goseq_down.png", width=18)
chr_uninf_down_msig_goseq$pvalue_plots$mfp_plot_over
dev.off()
```

```{r msig_goseq_sh}
ups <- pbmc_sig[["deseq"]][["ups"]][["sh_nil"]]
sh_uninf_msig_goseq <- goseq_msigdb(sig_genes=ups, signatures=broad_c7,
                                    signature_category="c7",
                                    length_db=hs_length,
                                    excel=glue::glue("excel/pbmc_sh_uninf_up_msig_goseq-v{ver}.xlsx"))
pp(file="images/pbmc_selfhealing_vs_uninfected_msig_goseq.png", width=18)
sh_uninf_msig_goseq$pvalue_plots$mfp_plot_over
dev.off()

downs <- pbmc_sig[["deseq"]][["downs"]][["sh_nil"]]
sh_uninf_down_msig_goseq <- goseq_msigdb(sig_genes=downs, signatures=broad_c7,
                                         signature_category="c7",
                                         length_db=hs_length,
                                         excel=glue::glue("excel/pbmc_sh_uninf_down_msig_goseq-v{ver}.xlsx"))
pp(file="images/pbmc_selfhealing_vs_uninfected_msig_goseq_down.png", width=18)
sh_uninf_down_msig_goseq$pvalue_plots$mfp_plot_over
dev.off()
```

```{r gsva_c8}
c8_pbmc_gsva <- simple_gsva(hs_cds_gsva_input,
                            signatures="reference/msigdb_v7.2/c8.all.v7.2.entrez.gmt",
                            signature_category="c8")
c8_pbmc_sig <- get_sig_gsva_categories(
    c8_pbmc_gsva,
    excel=glue::glue("excel/pbmc_batch_c8_categories-v{ver}.xlsx"))

pp(file="images/pbmc_c8_gsva_scores.svg")
c8_pbmc_sig$raw_plot
dev.off()
pp(file="images/pbmc_c8_likelihoods.svg")
c8_pbmc_sig$likelihood_plot
dev.off()
pp(file="images/pbmc_c8_subset.svg")
c8_pbmc_sig$subset_plot
dev.off()
```

# CIBERSORTx setup

## Signature matrix file

This is a matrix of expected gene abundances as rows and cell types as columns.
For the moment I am just going to use the example file:
"mixture_NSCLCbulk_Fig3g.txt".

## Mixture file

I think this is the exprs() data with hgnc IDs as rows.

```{r cibersort_mixture}
start <- exprs(hs_cds_norm)
xref <- fData(hs_cds_norm)[, c("ensembl_gene_id", "hgnc_symbol")]
new <- merge(xref, start, by.x="ensembl_gene_id", by.y="row.names")
rownames(new) <- make.names(new[["hgnc_symbol"]], unique=TRUE)
new[["hgnc_symbol"]] <- NULL
new[["ensembl_gene_id"]] <- NULL
readr::write_tsv(x=new, file="cibersort_mixture_file.txt")
```

## Merged class file

This makes 0 sense.

Maria Adelaida would like also to consider how the resolution is related to the
different donors.  Recall therefore that two donors are quite similar, and one
is rather different.  Thus, a differential expression analysis across donors
might prove helpful.

```{r multi_de_donors}
donor_cds_all <- set_expt_conditions(hs_cds_all, fact="donor")
donor_cds_minus <- set_expt_conditions(hs_cds_inf, fact="donor")
donor_cds_norm <- normalize_expt(donor_cds_all, filter=TRUE, transform="log2",
                                 convert="cpm", norm="quant")
plot_pca(donor_cds_norm)$plot
donor_cds_nb <- normalize_expt(donor_cds_all, filter=TRUE, transform="log2",
                               convert="cpm", batch="svaseq")
plot_pca(donor_cds_nb)$plot

donor_all_de <- all_pairwise(donor_cds_all, filter=TRUE, model_batch="svaseq")
donor_minus_de <- all_pairwise(donor_cds_minus, filter=TRUE, model_batch="svaseq")

donor_all_tables <- combine_de_tables(donor_all_de,
                                      excel="excel/all_donor_tables.xlsx")
donor_minus_tables <- combine_de_tables(donor_minus_de,
                                        excel="excel/minus_donor_tables.xlsx")
```

# Figure 4

Construct figure 4, this should include the following panels:

  a.  Library sizes of pbmc data
  b.  PCA of log2(quant(data)), with uninfected
  c.  PCA of log2(quant(data)), without uninfected
  d.  TSNE of b
  e.  TSNE of c

```{r figure_04}
pp(file=glue::glue("images/figure_4a-v{ver}.pdf"))
hs_inf_write$raw_libsize
dev.off()
pp(file=glue::glue("images/figure_4b-v{ver}.pdf"))
hs_inf_write$raw_scaled_pca
dev.off()
pp(file=glue::glue("images/figure_4c-v{ver}.pdf"))
hs_inf_write$norm_pca
dev.off()
```

```{r de}
keepers <- list("sh_nil" = c("sh", "uninf"),
                "ch_nil" = c("chr", "uninf"),
                "ch_sh" = c("chr", "sh"))
hs_pairwise_nobatch <- sm(all_pairwise(hs_cds_inf,
                                       model_batch=FALSE))
hs_pairwise_batch <- sm(all_pairwise(hs_cds_inf, model_batch=TRUE))
hs_pairwise_sva <- sm(all_pairwise(hs_cds_inf, model_batch="svaseq", filter=TRUE))
hs_nobatch_tables <- combine_de_tables(
    hs_pairwise_nobatch,
    keepers=keepers,
    excel=glue::glue("excel/hs_infect_nobatch-v{ver}.xlsx"))
hs_batch_tables <- combine_de_tables(
    hs_pairwise_batch,
    keepers=keepers,
    excel=glue::glue("excel/hs_infect_batch-v{ver}.xlsx"))
hs_sva_tables <- combine_de_tables(
    hs_pairwise_sva,
    keepers=keepers,
    excel=glue::glue("excel/hs_infect_sva-v{ver}.xlsx"))
```

## Extract the macrophage experiment

The following subset operation extract the samples used for the macrophage experiment. The next
three lines then change the colors from the defaults.

```{r macrophages}
new_colors <- c("#009900", "#990000", "#000099")
names(new_colors) <- c("uninf", "chr", "sh")

hs_macr <- subset_expt(hs_all, subset = "")
hs_macr <- set_expt_colors(hs_macr, colors=new_colors)
labels <- as.character(pData(hs_macr)[["label"]])
hs_macr <- set_expt_samplenames(hs_macr, labels)

hs_cds_macr <- set_expt_colors(hs_cds_macr, colors=new_colors)
hs_cds_macr <- set_expt_samplenames(hs_cds_macr, labels)
```

# Compare methods

In our meeting this morning (20190708), Najib and Maria Adelaida asked for how
the data looks depending on batch methodology.

```{r batch_methods}
start <- sm(normalize_expt(hs_cds_macr, filter=TRUE, convert="cpm",
                           norm="quant", transform="log2"))
pp(file="cpm_quant_filter_pca.png", image=plot_pca(start)$plot)

limma_batch <- sm(normalize_expt(hs_cds_macr, filter=TRUE,
                                 convert="cpm", norm="quant",
                                 transform="log2", batch="limma"))
pp(file="lcqf_limma.png", image=plot_pca(limma_batch)$plot)

pca_batch <- sm(normalize_expt(hs_cds_macr, filter=TRUE,
                               convert="cpm", norm="quant",
                               transform="log2", batch="pca"))
pp(file="lcqf_pca.png", image=plot_pca(pca_batch)$plot)

svaseq_batch <- sm(normalize_expt(hs_cds_macr, filter=TRUE,

                                  convert="cpm",
                                  transform="log2", batch="svaseq"))
pp(file="lcqf_svaseq.png", image=plot_pca(svaseq_batch)$plot)

sva_batch <- sm(normalize_expt(hs_cds_macr, filter=TRUE,
                               convert="cpm",
                               transform="log2", batch="fsva"))
pp(file="lcqf_sva.png", image=plot_pca(sva_batch)$plot)

combat_batch <- sm(normalize_expt(hs_cds_macr, filter=TRUE,
                                  convert="cpm",
                                  transform="log2", batch="combat"))
pp(file="lcqf_combat.png", image=plot_pca(combat_batch)$plot)

combatnp_batch <- sm(normalize_expt(hs_cds_macr, filter=TRUE,
                                    convert="cpm",
                                    transform="log2", batch="combat_noprior"))
pp(file="lcqf_combatnp.png", image=plot_pca(combatnp_batch)$plot)
```

# Figure S1

Figure S1 should include nice versions of the sample metrics.
The normalization chosen is batch-in-model.

First, however, we will make some plots of the raw data.

Sample names are going to be 'infectionstate_strainnumber' : chr_7721

* Panel A: Library sizes.
* Panel B: Heatmap distance raw.
* Panel C: PCA
* Panel D: TSNE

```{r fig_s1_write, fig.show="hide"}
fig_s1 <- sm(write_expt(
    hs_cds_macr, norm="raw", violin=FALSE, convert="cpm",
    transform="log2", batch="svaseq", filter=TRUE,
    excel=paste0("excel/figure_s1_sample_estimation-v", ver, ".xlsx")))

pp(file="fig_s1a_libsizes.tif", image=fig_s1$raw_libsize)
pp(file="fig_s1b_heatmap.tif", image=fig_s1$norm_disheat)
pp(file="fig_s1c_raw_pca.tif", image=fig_s1$raw_scaled_pca)
pp(file="fig_1a_norm_pca.tif", image=fig_s1$norm_pca)
```

# Differential Expression, Macrophage: `r ver`

# Differential expression analyses

The most likely batch method from the above is svaseq.  Let us perform
differential expression analyses with it along with a few other methods.

```{r setup_de_norm, fig.show="hide"}
hs_contrasts <- list(
    "macro_chr-sh" = c("chr","sh"),
    "macro_chr-nil" = c("chr","uninf"),
    "macro_sh-nil" = c("sh", "uninf"))
## Set up the data used in the comparative contrast sets.
```

## No batch in the model

### Set up no batch

Print a reminder of what we can expect when doing this with no batch information.

```{r nobatch_setup}
hs_macr_lowfilt <- sm(normalize_expt(hs_cds_macr, filter=TRUE))
hs_lowfilt_pca <- sm(plot_pca(hs_cds_macr, transform="log2"))
hs_lowfilt_pca$plot
```

```{r macro_nobatch1, fig.show="hide"}
hs_macr_nobatch <- all_pairwise(input=hs_cds_macr, model_batch=FALSE,
                                limma_method="robust")
## wow, all tools including basic agree almost completely
medians_by_condition <- hs_macr_nobatch$basic$medians
excel_file <- glue::glue("excel/{rundate}_hs_macr_nobatch_contr-v{ver}.xlsx")
hs_macr_nobatch_tables <- sm(combine_de_tables(hs_macr_nobatch,
                                               excel=excel_file,
                                               keepers=hs_contrasts,
                                               extra_annot=medians_by_condition))
excel_file <- glue::glue("excel/{rundate}_hs_macr_nobatch_sig-v{ver}.xlsx")
hs_macr_nobatch_sig <- sm(extract_significant_genes(hs_macr_nobatch_tables, lfc=0.585,
                                                    excel=excel_file,
                                                    according_to="all"))
```

## Batch in the model

### Batch setup

```{r batch_setup}
hs_lowfilt_batch_pca <- sm(plot_pca(hs_cds_macr, transform="log2",
                                    convert="cpm", batch="limma", filter=TRUE))
hs_lowfilt_batch_pca$plot
```

In this  attempt, we add a batch factor in the experimental model and see how it does.

```{r macro_batch1, fig.show="hide"}
## Here just let all_pairwise run on filtered data and do its normal ~ 0 + condition + batch analyses
hs_macr_batch <- all_pairwise(input=hs_cds_macr, batch=TRUE, limma_method="robust")
medians_by_condition <- hs_macr_batch$basic$medians
excel_file <- glue::glue("excel/{rundate}_hs_macr_batchmodel_contr-v{ver}.xlsx")
hs_macr_batch_tables <- combine_de_tables(
    hs_macr_batch,
    keepers=hs_contrasts,
    extra_annot=medians_by_condition,
    excel=excel_file)
excel_file <- glue::glue("excel/{rundate}_hs_macr_batchmodel_sig-v{ver}.xlsx")
hs_macr_batch_sig <- extract_significant_genes(
    hs_macr_batch_tables, excel=excel_file,  lfc=0.585,
    according_to="deseq")
excel_file <- glue::glue("excel/{rundate}_hs_macr_batchmodel_abund-v{ver}.xlsx")
hs_macr_batch_abun <- sm(extract_abundant_genes(
    hs_macr_batch_tables, excel=excel_file,
    according_to="deseq"))
```

## SVA

### sva setup

```{r sva_setup}
hs_lowfilt_sva_pca <- sm(plot_pca(hs_cds_macr, transform="log2",
                                  convert="cpm", batch="svaseq", filter=TRUE))
hs_lowfilt_sva_pca$plot
```

In this  attempt, we tell all pairwise to invoke svaseq.

```{r macro_sva, fig.show="hide"}
hs_macr_sva <- all_pairwise(input=hs_cds_macr, model_batch="svaseq", filter=TRUE, limma_method="robust")
medians_by_condition <- hs_macr_sva$basic$medians
excel_file <- glue::glue("excel/{rundate}_hs_macr_svamodel_contr-v{ver}.xlsx")
hs_macr_sva_tables <- combine_de_tables(
  hs_macr_sva,
  keepers=hs_contrasts,
  extra_annot=medians_by_condition,
  excel=excel_file)
excel_file <- glue::glue("excel/{rundate}_hs_macr_svamodel_sig-v{ver}.xlsx")
hs_macr_sva_sig <- extract_significant_genes(
    hs_macr_sva_tables, excel=excel_file,
    according_to="deseq")
excel_file <- glue::glue("excel/{rundate}_hs_macr_svamodel_abund-v{ver}.xlsx")
hs_macr_sva_abun <- sm(extract_abundant_genes(
    hs_macr_sva_tables, excel=excel_file,
    according_to="deseq"))
```

## Combat

As you know, combat actually changes the data, so it will require a slightly
different setup.

### combat setup

```{r combat_setup}
combatnp_input <- normalize_expt(hs_cds_macr, filter=TRUE,
                                 convert="cpm",
                                 batch="combat_noprior")
```

In this  attempt, we tell all pairwise to invoke svaseq.

```{r macro_combat, fig.show="hide"}
hs_macr_combat <- all_pairwise(input=combatnp_input, model_batch=FALSE,
                               force=TRUE, limma_method="robust")
excel_file <- glue::glue("excel/{rundate}_hs_macr_combat_contr-v{ver}.xlsx")
hs_macr_combat_tables <- combine_de_tables(
    hs_macr_combat,
    keepers=hs_contrasts,
    extra_annot=medians_by_condition,
    excel=excel_file)
excel_file <- glue::glue("excel/{rundate}_hs_macr_combat_sig-v{ver}.xlsx")
hs_macr_combat_sig <- extract_significant_genes(
  hs_macr_combat_tables, excel=excel_file,
  according_to="deseq")
excel_file <- glue::glue("excel/{rundate}_hs_macr_combat_abund-v{ver}.xlsx")
hs_macr_combat_abun <- sm(extract_abundant_genes(
  hs_macr_combat_tables, excel=excel_file,
  according_to="deseq"))
```

## Compare these three results

```{r compare_de_results}
nobatch_batch <- compare_de_results(first=hs_macr_nobatch_tables,
                                    second=hs_macr_batch_tables)
nobatch_batch$logfc

batch_sva <- compare_de_results(first=hs_macr_batch_tables,
                                second=hs_macr_sva_tables)
batch_sva$logfc

batch_combat <- compare_de_results(first=hs_macr_batch_tables,
                                   second=hs_macr_combat_tables)
batch_combat$logfc

sva_combat <- compare_de_results(first=hs_macr_sva_tables,
                                 second=hs_macr_combat_tables)
sva_combat$logfc
## Interesting how much sva and combat disagree.
```

## Two likely volcano plots

# TODO 202007

1.  Change volcano plots to logFC 0.58.
2.  Send figure S1 new versions with improved labels.

```{r table_s2}
batchmodel_volcano <- plot_volcano_de(
    table=hs_macr_batch_tables[["data"]][["macro_chr-sh"]],
    color_by="state",
    fc_col="deseq_logfc",
    p_col="deseq_adjp",
    logfc=0.58,
    shapes_by_state=FALSE,
    line_position="top")
batchmodel_volcano$plot

svamodel_volcano <- plot_volcano_de(
    table=hs_macr_sva_tables[["data"]][["macro_chr-sh"]],
    color_by="state",
    fc_col="deseq_logfc",
    p_col="deseq_adjp",
    logfc=0.58,
    shapes_by_state=FALSE,
    line_position="top")
pp(file="sva_chsh_deseq_volcano.tif", image=svamodel_volcano$plot)
```

## PROPER

```{r proper}
hs_proper <- simple_proper(de_tables=hs_macr_batch_tables)
```

## Ontology searching against the sva results.

Recall that I made the variables 'hs_macr_sva_sig' and 'hs_macr_sva_abun' to
hold the results of the most significantly changed and abundant genes.

gProfiler is my favorite tool for ontology searching, however they recently had
a big update and split their code.  The new version has all sorts of cool toys,
but as of the last time I tried it, did not work.  Thus the following is still
using the older methods.

```{r ontology2}
lfs_up <- hs_macr_sva_sig[["deseq"]][["ups"]][["macro_chr-sh"]]
lfs_down <- hs_macr_sva_sig[["deseq"]][["downs"]][["macro_chr-sh"]]

up_gp <- simple_gprofiler(sig_genes=lfs_up, species="hsapiens")
up_gp[["pvalue_plots"]][["bpp_plot_over"]]
down_gp <- simple_gprofiler(sig_genes=lfs_down, species="hsapiens")
down_gp[["pvalue_plots"]][["bpp_plot_over"]]

up_goseq <- simple_goseq(sig_genes=lfs_up, go_db=hs_go[["go"]], length_db=hs_lengths)
up_goseq[["pvalue_plots"]][["bpp_plot_over"]]
down_goseq <- simple_goseq(sig_genes=lfs_down, go_db=hs_go[["go"]], length_db=hs_lengths)
down_goseq[["pvalue_plots"]][["bpp_plot_over"]]

up_topgo <- simple_topgo(sig_genes=lfs_up, go_db=hs_go[["go"]])
up_topgo[["pvalue_plots"]][["bpp_plot_over"]]
down_topgo <- simple_topgo(sig_genes=lfs_down, go_db=hs_go[["go"]])
down_topgo[["pvalue_plots"]][["bpp_plot_over"]]

up_cp <- simple_clusterprofiler(sig_genes=lfs_up, do_david=FALSE, do_gsea=FALSE, orgdb="org.Hs.eg.db", fc_column="deseq_logfc")
up_cp[["pvalue_plots"]][["ego_all_bp"]]
up_cp[["plots"]][["dot_all_bp"]]
down_cp <- simple_clusterprofiler(sig_genes=lfs_down, do_david=FALSE, do_gsea=FALSE, orgdb="org.Hs.eg.db", fc_column="deseq_logfc")
down_cp[["pvalue_plots"]][["ego_all_bp"]]
down_cp[["plots"]][["dot_all_bp"]]
```

# Separate the three donors

```{r separate_donors}
all_de <- all_pairwise(hs_cds_all, model_batch="svaseq", filter=TRUE)

d108 <- subset_expt(hs_cds_all, subset="donor=='d108'")
d107 <- subset_expt(hs_cds_all, subset="donor=='d107'")
d110 <- subset_expt(hs_cds_all, subset="donor=='d110'")

d108_de <- all_pairwise(d108, model_batch="svaseq", filter=TRUE)
d107_de <- all_pairwise(d107, model_batch="svaseq", filter=TRUE)
d110_de <- all_pairwise(d110, model_batch="svaseq", filter=TRUE)

d108_table <- combine_de_tables(d108_de, excel="excel/d108_tables.xlsx")
d107_table <- combine_de_tables(d107_de, excel="excel/d107_tables.xlsx")
d110_table <- combine_de_tables(d110_de, excel="excel/d110_tables.xlsx")


d108 <- subset_expt(hs_cds_inf, subset="donor=='d108'")
d107 <- subset_expt(hs_cds_inf, subset="donor=='d107'")
d110 <- subset_expt(hs_cds_inf, subset="donor=='d110'")
d108_de <- all_pairwise(d108, model_batch="svaseq", filter=TRUE)
d107_de <- all_pairwise(d107, model_batch="svaseq", filter=TRUE)
d110_de <- all_pairwise(d110, model_batch="svaseq", filter=TRUE)
d108_table <- combine_de_tables(d108_de, excel="excel/d108_minustwo_tables.xlsx")
d107_table <- combine_de_tables(d107_de, excel="excel/d107_minustwo_tables.xlsx")
d110_table <- combine_de_tables(d110_de, excel="excel/d110_minustwo_tables.xlsx")
```

```{r saveme}
pander::pander(sessionInfo())
message(paste0("This is hpgltools commit: ", get_git_commit()))
this_save <- paste0(gsub(pattern="\\.Rmd", replace="", x=rmd_file), "-v", ver, ".rda.xz")
message(paste0("Saving to ", this_save))
tmp <- sm(saveme(filename=this_save))
```


```{r loadme, eval=FALSE, include=FALSE}
this_save <- paste0(gsub(pattern="\\.Rmd", replace="", x=rmd_file), "-v", ver, ".rda.xz")
loaded <- loadme(filename=this_save)
```
